### PR TITLE
added OpenSSL::Digest

### PR DESCRIPTION
**Pull request by benoist**

### CODE REVIEW

File: src/openssl/digest/digest_base.cr [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    def <<(data)
+      update(data)
+    end
+
+    def base64digest
+      Base64.encode(digest)
+    end
+
+    def hexdigest
+      DigestBase.hexdump(digest)
```
Comment (ysbaddaden) [inline src/openssl/digest/digest_base.cr]: I don't think we need a custom `hexdump`, `digest.hexstring` should do the job.

File: src/openssl/digest/digest.cr [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    def update(data : String | Slice)
+      LibCrypto.evp_digestupdate(self, data, data.size)
+      self
+    end
+
+    protected def finish
+      size = digest_size
+      data = Slice(UInt8).new(size)
+      LibCrypto.evp_digestfinal_ex(@ctx, data, nil)
+      data
```
Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: This should be faster:

```crystal
ptr = Pointer(UInt8).new(size)
LibCrypto.evp_digestfinal_ex(@ctx, ptr, nil)
ptr.to_slice(size)
```

File: src/openssl/digest/digest.cr [hunk 3]
```diff
@@ -0,0 +1,88 @@
+require "../lib_crypto"
+require "./digest_base"
+
+module OpenSSL
+  class Digest
+    class DigestError < OpenSSL::Error; end
```
Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: `OpenSSL::Digest::Error`

File: src/openssl/digest/digest.cr [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    protected def self.create_evp_mt_ctx(name)
+      md = LibCrypto.evp_get_digestbyname(name)
+      unless md
+        oid = LibCrypto.obj_txt2obj(name, 0)
+        md = LibCrypto.evp_get_digestbyname(LibCrypto.obj_nid2sn(LibCrypto.obj_obj2nid(oid)))
+        LibCrypto.asn1_object_free(oid)
+      end
+      unless md
+        raise "Unsupported digest algoritm: #{name}"
```
Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: Please raise an `OpenSSL::Digest::Error`.

File: src/openssl/digest/digest.cr [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        raise OpenSSL::Digest::DigestError.new "Digest initialization failed."
+      end
+      if LibCrypto.evp_digestinit_ex(ctx, md, nil) != 1
+        raise OpenSSL::Digest::DigestError.new "Digest initialization failed."
+      end
+      ctx
+    end
+
+    def self.new(name)
+      new(name, create_evp_mt_ctx(name))
```
Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: Could you add specs for the different cases in `create_evp_mt_ctx`? The digest is unsupported, get digest by OID, etc.

File: src/openssl/digest/digest.cr (hunk 2 above)
Comment (asterite) [inline src/openssl/digest/digest.cr]: I think you meant `Pointer.malloc(size)`, `Pointer.new(size)` just creates one with that address

Comment (technorama) [inline src/openssl/digest/digest.cr]: Why would that be faster?  The allocation code seems almost identical.  `to_slice` calls `Slice.new`.

Comment (benoist) [inline src/openssl/digest/digest.cr]: I'll test both versions :-)

Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: My bad. I got mistaken by `Pointer(T).new` which is a lot faster... since it doesn't malloc.

Comment (benoist) [inline src/openssl/digest/digest.cr]:
```
benoist@benoist-i7$ ./bin/crystal run worksheet.cr --release
  slice 635.21k (± 0.59%)  1.02× slower
pointer 650.09k (± 1.89%)       fastest
[~/code/crystal] feature/openssl-digest
benoist@benoist-i7$ ./bin/crystal run worksheet.cr --release
  slice 587.44k (± 1.75%)  1.07× slower
pointer 630.61k (± 0.62%)       fastest
[~/code/crystal] feature/openssl-digest
benoist@benoist-i7$ ./bin/crystal run worksheet.cr --release
  slice 597.98k (± 0.46%)       fastest
pointer 596.82k (± 0.68%)  1.00× slower
```

Not a big difference

File: src/openssl/digest/digest.cr (hunk 5 above)
Comment (benoist) [inline src/openssl/digest/digest.cr]: Don't know how to trigger any of the other errors, but I've added spec for an unsupported digest

Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: No problem for the error checks. But it woud be nice to test this part:

```crystal
unless md
  oid = LibCrypto.obj_txt2obj(name, 0)
  md = LibCrypto.evp_get_digestbyname(LibCrypto.obj_nid2sn(LibCrypto.obj_obj2nid(oid)))
  LibCrypto.asn1_object_free(oid)
end
```

Comment (benoist) [inline src/openssl/digest/digest.cr]:
```crystal
md = LibCrypto.evp_get_digestbyname(name)
unless md
  LibCrypto.obj_txt2obj(name, 0)
```
@datanoise @technorama @ysbaddaden i'm not sure when md will be nil as both functions require a string. Any suggestions?

Comment (technorama) [inline src/openssl/digest/digest.cr]: @ysbaddaden I don't see the need for additional error checking.  `oid` could be `NID_undef` but I think that case is handled a few lines down when checking if `md` is `nil`.  Did I miss something?

Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: Isn't it trying to load a digest by oid when loading by name failed in the outlined part? I don't know if it's any useful, thought.

Comment (benoist) [inline src/openssl/digest/digest.cr]: @ysbaddaden I've tried both methods and it looks to me like the both yield the same result with the same name input. I don't know if different versions of OpenSSL will react the same way. So perhaps we can just remove it for now and see if it causes any problems?

Comment (ysbaddaden) [inline src/openssl/digest/digest.cr]: OK.

Comment (benoist) [inline src/openssl/digest/digest.cr]: Done :)

**Comment by ysbaddaden** [comment]
Let's merge!

**Comment by RX14** [comment]
Anyone know what the reasoning behind splitting up Digest and DigestBase is? DigestBase doesn't appear to be included anywhere else.

**Comment by benoist** [comment]
if I remember correctly the original code from openssl.cr shard contained separate classes for each digest. As not every version of OpenSSL supports the same digest, this was removed, but the base was never included in the current location.

**Pull request merged**